pennylane/operation.py

```diff
@@ -695,5 +742,32 @@ def heisenberg_obs(self, num_wires):
             array[float]: :math:`q`
         """
         p = self.parameters
-        U = self._heisenberg_rep(p) # pylint: disable=assignment-from-none
+        U = self._heisenberg_rep(p)  # pylint: disable=assignment-from-none
         return self.heisenberg_expand(U, num_wires)
+
+
+class Tensor:
+    """Tensor wrapper for operations and observables.
+
+    :class:`Tensor` is used to wrap tensors of observables.
+
+    Args:
+        ops (list of Observable): Quantum observable or a list of observables.
+    """
+
+    # pylint: disable=abstract-method
+    return_type = None
+
+    def __init__(self):
+        self.ops = []
+        self.params = None
+        self.wires = []
+        self.name = "Tensor"
+        self.parameters = None
+
+    def append(self, op):
+        """Add operations to a tensor list
+        """
+        self.ops.append(op)
+        for w in op.wires:
+            self.wires.append(w)
```

Comment:
I just had a thought - do we need a separate Tensor class? Perhaps its functionality could be rolled directly into the `Observable` class, by adding the method
```python
    def append(self, op):
        """Add operations to a tensor list"""
        self.ops.append(op)
        self.name.append(op.name)
        self.wires.append(op.wires)
        self.params.append(op.params)
        return self

    __iadd__ = append
    __add__ = append
```

This way, to define a Tensor operation, you could simply do
```python
tensor_obs = PauliX(0) + Hermitian(A, wires=[1,2])
```
internally within the `expval`, `var`, or `sample` functions:
```python
if len(ops) > 1:
    # tensor of observables
    tensor_ops = ops[0]
    for o in ops:
        tensor_ops += o
    tensor_ops.return_type = "expectation"
````

Or perhaps I'm missing something that would stop this from working?

Reply:
@josh146 maybe from a user-interface perspective that's not to best way to do things? If there was a clear operator we could overload, then it makes sense, but to me `+` immediately brings to mind something like
X_0\otimes\Identity_1 + Hermitian_{0,1}

Reply:
Alternate suggestions: 
```PauliX(0) @ Hermitian(A, wires=[0,1])``` <-- this one seems most intuitive
```PauliX(0) % Hermitian(A, wires=[0,1])```

Reply:
Note that this won't be user facing, just used internally (so perhaps we shouldn't be operator overloading at all and just using the method 😆)

Instead, the user will be using the `expval(PauliX(0), PauliY(1))` syntax.

Reply:
If it's useful, why not make it user-facing?

```
big_obs = PauliX(0) @ PauliY(1) @ PauliZ(2)
def circuit1(pars):
   ...
   return qml.expval(big_obs)

def circuit1(pars):
   ...
   return qml.sample(big_obs)
```
etc.

Reply:
I don't mind that! it would have to be done from within the QNode though so that queueing can take place:

```python
def circuit1(pars):
   big_obs = PauliX(0) @ PauliY(1) @ PauliZ(2)
   return qml.expval(big_obs)
```

(More reason to try and solve/replace queuing.)

My only reason to slightly preferring `expval(PauliX(0), PauliY(1))` over `expval(PauliX(0) @ PauliY(1))` is just code complexity. The latter simply requires `*args` in the expval signature, whereas operator overloading requires new methods and could result in more edge-cases to test.

Reply:
Even if this is a little bit old now (mabye you agreed on something in the meantime), I'd like to add that I really don't like `@` as an operator for tensor products because it clearly has the meaning of the matrix product in python and measuring `X @ H` could be mistaken as measuring the product of X and H and not the tensor product.

What would be possible is to construct [infix operators](http://tomerfiliba.com/blog/Infix-Operators/):
```python
qml.expval(qml.PauliX(0) |tensor| qml.PauliY(1) |tensor| qml.Hermitian(A, [3, 4]))
qml.expval(qml.PauliX(0) |x| qml.PauliY(1) |x| qml.Hermitian(A, [3, 4]))
qml.expval(qml.PauliX(0) <<x>> qml.PauliY(1) <<x>> qml.Hermitian(A, [3, 4]))
```
But I don't thinkt that would be a good idea because this will introduce a lot of complexity without much improvement over `Tensor(qml.PauliX(0), qml.PauliY(1), qml.Hermitian(A, [3, 4]))`